src/common/bitfield_util.h

```diff
@@ -0,0 +1,180 @@
+    }
+    return 1 <= bits && bits <= max_bits;
+  }
+
+  BitfieldEncoding(Type type, uint8_t bits) : type_(type), bits_(bits) {
+    if (!IsSupportedBitLengths(type, bits)) {
+      throw BitfieldBitsLengthNotFitExecption();
+    }
+  }
+
+  bool IsSigned() const noexcept { return type_ == Type::kSigned; }
+
+  uint8_t Bits() const noexcept { return bits_; }
+
+  void SetBits(uint8_t new_bits) {
+    if (!IsSupportedBitLengths(type_, new_bits)) {
+      throw BitfieldBitsLengthNotFitExecption();
+    }
+    bits_ = new_bits;
+  }
+
+  void SetType(Type new_type) {
+    if (!IsSupportedBitLengths(new_type, bits_)) {
+      throw BitfieldBitsLengthNotFitExecption();
+    }
+    type_ = new_type;
+  }
+
+  std::string ToString() const noexcept { return (IsSigned() ? "i" : "u") + std::to_string(static_cast<int>(bits_)); }
+
+ private:
+  Type type_;
+  uint8_t bits_;
+};
+
+struct BitfieldOperation {
+  // see https://redis.io/commands/bitfield/ to get more details.
+  enum class Type : uint8_t { kGet, kSet, kIncrBy };
+
+  Type type;
+  BitfieldOverflowBehavior overflow{BitfieldOverflowBehavior::kWrap};
+  BitfieldEncoding encoding{BitfieldEncoding::Type::kSigned, 32};
+  uint32_t offset;
+  // INCRBY amount or SET value
+  int64_t value;
+};
+
+namespace detail {
+// return true if overflow
+bool SignedBitfieldPlus(uint64_t value, int64_t incr, uint8_t bits, BitfieldOverflowBehavior overflow, uint64_t *dst);
+
+// return true if overflow.
+bool UnsignedBitfieldPlus(uint64_t value, int64_t incr, uint8_t bits, BitfieldOverflowBehavior overflow, uint64_t *dst);
+}  // namespace detail
+
+// safe cast from unsigned to signed, without any bit changes.
+// see also "Integral conversions" on https://en.cppreference.com/w/cpp/language/implicit_conversion
+// If the destination type is signed, the result when overflow is implementation-defined until C++20
+inline int64_t CastToSignedWithoutBitChanges(uint64_t x) {
+  int64_t res = 0;
+  memcpy(&res, &x, sizeof(res));
+  return res;
+}
+
+// return true if overflow.
+inline auto BitfieldPlus(uint64_t value, int64_t incr, BitfieldEncoding enc, BitfieldOverflowBehavior overflow,
+                         uint64_t *dst) {
+  if (enc.IsSigned()) {
+    return detail::SignedBitfieldPlus(value, incr, enc.Bits(), overflow, dst);
+  }
+  return detail::UnsignedBitfieldPlus(value, incr, enc.Bits(), overflow, dst);
+}
+
+// return true if successful
+inline bool BitfieldOp(BitfieldOperation op, uint64_t old_value, uint64_t *new_value) {
+  if (op.type == BitfieldOperation::Type::kGet) {
+    *new_value = old_value;
+    return true;
+  }
+
+  bool overflow = false;
+  if (op.type == BitfieldOperation::Type::kSet) {
+    overflow = BitfieldPlus(op.value, 0, op.encoding, op.overflow, new_value);
+  } else {
+    overflow = BitfieldPlus(old_value, op.value, op.encoding, op.overflow, new_value);
+  }
+
+  return op.overflow != BitfieldOverflowBehavior::kFail || !overflow;
+}
+
+template <class BitmapView>
```

Comment:
We can leave a todo here for optimizations. I didn't try it in `-O3` but it looks slow if bits isn't small